### PR TITLE
Check access token scope

### PR DESCRIPTION
**Pull request by danlee1025**
- Check access token scope before authorizing request
 - Added `system/*.*` and `system/*.write` scopes in addition to `system/*.read`
     - All requests are allowed when scope includes `system/*.*`
     - GET requests are allowed when scope includes `system/*.read`
     - POST, PUT, and DELETE requests are allowed when scope includes `system/*.write`
 - To test, you must first get an access token from the keycloak server. I added another client with the client_id `test` which made scopes optional. client_secret can be found on the keycloak server.

### CODE REVIEW

**Review by dehall — APPROVED** [review]
This looks good. We could theoretically make it more generic by allowing users of the library to pass in more detailed logic (ex, filtering by resource type, etc), but that's probably overkill.

File: src/main/java/org/mitre/hapifhir/BackendAuthorizationInterceptor.java [hunk 1]
```diff
@@ -116,15 +130,40 @@ private RSAPublicKey getKey()
...
+    /**
+     * Allow any request if scope includes "system/*.*"
+     * Allow GET request if scope includes "system/*.read"
+     * Allow POST, PUT, DELETE requests if scope includes "system/*.write"
+     */
+    String requestType = theRequestDetails.getRequestType().name();
+    if (scopes.contains("system/*.*") ||
+        (requestType == "GET" && scopes.contains("system/*.read")) ||
+        ((requestType == "POST" || requestType == "PUT" || requestType == "DELETE") &&
+        scopes.contains("system/*.write"))) return;
```
Comment (blangley28) [inline src/main/java/org/mitre/hapifhir/BackendAuthorizationInterceptor.java]: In Java String is an object so `==` checks address not data content. I suggest just using the `RequestTypeEnum` here anyway
```suggestion
    RequestTypeEnum requestType = theRequestDetails.getRequestType();
    if (scopes.contains("system/*.*") ||
        (requestType.equals(RequestTypeEnum.GET) && scopes.contains("system/*.read")) ||
        ((requestType.equals(RequestTypeEnum.POST) || requestType.equals(RequestTypeEnum.PUT) || requestType.equals(RequestTypeEnum.DELETE)) &&
        scopes.contains("system/*.write"))) return;
```

**Review by blangley28 — CHANGES_REQUESTED** [review]
Hopefully someone else can comment on this but is there any standard for JWT access tokens? We removed the dependency on Keycloak to make this work for any JWT access token but we are now expecting the JWT to have a `scope` claim. Obviously Keycloak includes this, which is why we are getting reading that claim here, but it makes me question. We should either:
A) identify a standard JWT access token structure this follows and point it out in the README
B) explicitly list any assumptions of the JWT structure in the README so users know their access token MUST include a scope claim for this library to work

File: src/main/java/org/mitre/hapifhir/BackendAuthorizationInterceptor.java (hunk 1 above)
Comment (danlee1025) [reply src/main/java/org/mitre/hapifhir/BackendAuthorizationInterceptor.java]: Should be fixed.

**Comment by danlee1025** [comment]
JWT does have a [specification](https://tools.ietf.org/html/rfc7519). Our scopes are listed in the well known endpoint and a user uses an access token without scope, they'll get an invalid scope error. We could add something to the README to clarify though.

**Comment by dehall** [comment]
Not sure about JWTs more broadly, but since we're following the SMART backend services auth guide a scope is expected at the top level: http://build.fhir.org/ig/HL7/bulk-data-export/authorization/index.html#issuing-access-tokens.
We could include a link to that page in the README

**Comment by blangley28** [comment]
@dehall the link you provided is for the JSON returned from the `/token` request, not the structure of the token itself. AFAIK SMART does not define any restrictions on the access token (doesn't even need to be a JWT)

**Comment by blangley28** [comment]
@danlee1025 doesn't look like the JWT specification includes `scope` at all. So it seems like this is an implementation specific details which should be documented

**Comment by dehall** [comment]
Ah I see
>  The method the server uses to validate the access token is beyond the scope of this specification but generally involves an interaction or coordination between the resource server and the authorization server.

Ugh. Maybe we need to implement calling out the token introspection URL as well. https://tools.ietf.org/html/rfc7662

**Comment by blangley28** [comment]
That doesn't seem like a secure way to determine the scope though. Could easily be subjected to MIM

**Comment by dehall** [comment]
How? The resource server knows and should be able to trust the authentication server. It would be a fixed url like the certs/token url. Only the content of the request would vary based on what a client sent as a token.

**Comment by blangley28** [comment]
Wouldn't this server need to have an established secret with (aka check the SSL certs of the auth server itself then)? Couldn't someone intercept the request and send a response pretending to be the auth server. So we would need to verify the response is indeed from the auth server

**Comment by dehall** [comment]
Yes, and unfortunately that's out of scope of the spec too.
> If left unprotected and un-throttled, the introspection endpoint
   could present a means for an attacker to poll a series of possible
   token values, fishing for a valid token.  To prevent this, the
   authorization server MUST require authentication of protected
   resources that need to access the introspection endpoint and SHOULD
   require protected resources to be specifically authorized to call the
   introspection endpoint.  The specifics of such authentication
   credentials are out of scope of this specification, but commonly
   these credentials could take the form of any valid client
   authentication mechanism used with the token endpoint, an OAuth 2.0
   access token, or other HTTP authorization or authentication
   mechanism.

**Review by blangley28 — APPROVED** [review]

**Pull request merged**